### PR TITLE
[WIP] Trace events writer

### PR DESCRIPTION
**Pull request by vandanavk**
### Description of changes:
This PR adds a trace event writer to smdebug along with tests.

List of tests:
1. Event file creation and checking validity
2. Multiple processes writing to multiple files
3. Testing file size and file interval rotation policy
4. Testing duration events
5. Testing creation of timeline event using MXNet hook
6. Recording trace events in PT distributed training
7. Recording trace events in TF mirrored strategy - TODO (need to annotate keras.py)
8. Recording trace events in TF eager and non-eager mode - TODO

### Test results:
1. `test_rotation_policy(file_interval=1 second)`

/tmp/test/framework/pevents/20051520/1589600102_36468_8c859046be41.ant.amazon.com_pythontimeline.json
```
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589600101948949}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "RotationPolicyTest_file_interval"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event1", "pid": 1, "ph": "X", "ts": 1004767, "dur": 1000000, "args": {"step_num": 1}},
{"name": "event2", "pid": 1, "ph": "X", "ts": 2009510, "dur": 1000000, "args": {"step_num": 2}}
]
```
/tmp/test/framework/pevents/20051520/1589600104_36468_8c859046be41.ant.amazon.com_pythontimeline.json
```
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589600103961304}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "RotationPolicyTest_file_interval"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event3", "pid": 1, "ph": "X", "ts": 1000477, "dur": 1000000, "args": {"step_num": 3}},
{"name": "event4", "pid": 1, "ph": "X", "ts": 2004723, "dur": 1000000, "args": {"step_num": 4}}
]
```
2. `test_rotation_policy(file_size=300 bytes)`
/tmp/test/framework/pevents/20051520/1589600651_49118_8c859046be41.ant.amazon.com_pythontimeline.json
```
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589600650503274}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "RotationPolicyTest_file_size"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event1", "pid": 1, "ph": "X", "ts": 1003807, "dur": 1000000, "args": {"step_num": 1}}
]
```
/tmp/test/framework/pevents/20051520/1589600652_49118_8c859046be41.ant.amazon.com_pythontimeline.json
```
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589600651508992}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "RotationPolicyTest_file_size"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event2", "pid": 1, "ph": "X", "ts": 1001307, "dur": 1000000, "args": {"step_num": 2}}
]
```
/tmp/test/framework/pevents/20051520/1589600653_49118_8c859046be41.ant.amazon.com_pythontimeline.json
```
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589600652513364}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "RotationPolicyTest_file_size"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event3", "pid": 1, "ph": "X", "ts": 1002049, "dur": 1000000, "args": {"step_num": 3}}
]
```
/tmp/test/framework/pevents/20051520/1589600654_49118_8c859046be41.ant.amazon.com_pythontimeline.json
```
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589600653518509}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "RotationPolicyTest_file_size"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event4", "pid": 1, "ph": "X", "ts": 1000750, "dur": 1000000, "args": {"step_num": 4}}
]
```
3. test_duration_events (total 10 events captured. showing only 5 below)
```
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589600835227643}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "DurationEventTest"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event1", "pid": 1, "ph": "B", "ts": 5084, "args": {"step_num": 1}},
{"name": "event1", "pid": 1, "ph": "E", "ts": 5136, "args": {"step_num": 1}},
{"name": "event2", "pid": 1, "ph": "B", "ts": 5155, "args": {"step_num": 2}},
{"name": "event2", "pid": 1, "ph": "E", "ts": 5165, "args": {"step_num": 2}},
{"name": "event3", "pid": 1, "ph": "B", "ts": 5175, "args": {"step_num": 3}},
{"name": "event3", "pid": 1, "ph": "E", "ts": 5184, "args": {"step_num": 3}},
{"name": "event4", "pid": 1, "ph": "B", "ts": 5192, "args": {"step_num": 4}},
{"name": "event4", "pid": 1, "ph": "E", "ts": 5200, "args": {"step_num": 4}},
{"name": "event5", "pid": 1, "ph": "B", "ts": 5208, "args": {"step_num": 5}},
{"name": "event5", "pid": 1, "ph": "E", "ts": 5216, "args": {"step_num": 5}},
```
4. test_create_timeline_file
/tmp/test/framework/pevents/20051520/1589601087_61797_8c859046be41.ant.amazon.com_pythontimeline.json
```
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589601087300287}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "FileCreationTest"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event1", "pid": 1, "ph": "X", "ts": 2133, "dur": 1000000, "args": {"step_num": 1}},
{"name": "event2", "pid": 1, "ph": "X", "ts": 2153, "dur": 1000000, "args": {"step_num": 2}},
{"name": "event3", "pid": 1, "ph": "X", "ts": 2247, "dur": 1000000, "args": {"step_num": 3}},
{"name": "event4", "pid": 1, "ph": "X", "ts": 2260, "dur": 1000000, "args": {"step_num": 4}},
{"name": "event5", "pid": 1, "ph": "X", "ts": 2310, "dur": 1000000, "args": {"step_num": 5}},
{"name": "event6", "pid": 1, "ph": "X", "ts": 2316, "dur": 1000000, "args": {"step_num": 6}},
{"name": "event7", "pid": 1, "ph": "X", "ts": 2323, "dur": 1000000, "args": {"step_num": 7}},
{"name": "event8", "pid": 1, "ph": "X", "ts": 2326, "dur": 1000000, "args": {"step_num": 8}},
{"name": "event9", "pid": 1, "ph": "X", "ts": 2332, "dur": 1000000, "args": {"step_num": 9}},
{"name": "event10", "pid": 1, "ph": "X", "ts": 2336, "dur": 1000000, "args": {"step_num": 10}}
]
```

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by vandanavk** [comment]
When multiple processes try to write to the json file, the output looks like this. 

```
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589381941318049}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "MultiProcessTest"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event1", "ph": "X", "ts": 16662, "pid": 1, "dur": 100000, "args": {"step number": 0, "process_rank": 0}},
{"name": "event2", "ph": "X", "ts": 16798, "pid": 1, "dur": 100000, "args": {"step number": 1}},
[
{"name": "process_name", "ph": "M", "pid": 0, "args": {"name": "start_time_since_epoch_in_micros", "value": 1589381941318049}},
{"name": "process_sort_index", "ph": "M", "pid": 0, "args": {"sort_index": 0}},
{"name": "process_name", "ph": "M", "pid": 1, "args": {"name": "MultiProcessTest"}},
{"name": "process_sort_index", "ph": "M", "pid": 1, "args": {"sort_index": 1}},
{"name": "event1", "ph": "X", "ts": 21228, "pid": 1, "dur": 100000, "args": {"step number": 0, "process_rank": 1}},
{"name": "event2", "ph": "X", "ts": 21408, "pid": 1, "dur": 100000, "args": {"step number": 1}},
```

File: smdebug/core/locations.py [hunk 1]
```diff
@@ -105,6 +106,15 @@ def get_file_location(self, base_dir=""):
         return os.path.join(event_key_prefix, self.get_filename())
 
 
+class TraceFileLocation:
+    def __init__(self, mode=None):
+        self.mode = mode
+
+    @property
+    def get_file_location(self):
+        return os.getenv(SM_PROFILER_FILE_PATH_ENV_STR, DEFAULT_SAGEMAKER_PROFILER_PATH)
```
Comment (Vikas-kum) [inline smdebug/core/locations.py:115]: get file location return is profiler config json?

File: smdebug/core/writer.py [hunk 2]
```diff
@@ -90,17 +96,24 @@ def __init__(
             )
             event_file_path = el.get_file_location(base_dir=self.trial_dir)
             self.index_writer = None
+        elif wtype == "trace":
+            el = TraceFileLocation(mode=self.mode)
```
Comment (Vikas-kum) [inline smdebug/core/writer.py:100]: mode ?

File: smdebug/core/tfevent/timeline_file_writer.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+from smdebug.core.access_layer.file import TSAccessFile
+from smdebug.core.access_layer.s3 import TSAccessS3
+from smdebug.core.utils import is_s3
+
+
+class TimelineWriter(object):
+    def __init__(self, file_path):
+        """ Writer is initialized upon adding the first index. """
+        self.file_path = file_path
+        self.event_payload = []
```
Comment (Vikas-kum) [inline smdebug/core/tfevent/timeline_file_writer.py:16]: event_pay_load?

File: smdebug/core/hook.py [hunk 4]
```diff
@@ -477,6 +484,21 @@ def _maybe_get_tb_writer(self) -> Optional[FileWriter]:
...
 
+    def _maybe_get_timeline_writer(self) -> Optional[FileWriter]:
+        """ Returns a FileWriter object if timeline_writer has been created, else creates a file at the
+        location specified by the environment variable SM_PROFILER_FILE_PATH and returns the FileWriter.
+
+        """
+        if not self.timeline_writer:
+            self.timeline_writer = FileWriter(
+                trial_dir=self.out_dir,
+                step=self.step,
```
[the remark this answers is not in the document]
Comment (vandanavk) [reply smdebug/core/hook.py]: not relevant in latest commit

File: smdebug/core/hook.py [hunk 5]
```diff
@@ -624,6 +650,17 @@ def _write_histogram_summary(self, tensor_name, tensor_value, save_collections):
                     )
                     break
 
+    def _write_trace_event_summary(self, tensor_name, duration, timestamp, step_num, worker):
+        timeline_writer = self._maybe_get_timeline_writer()
+        if timeline_writer:
+            timeline_writer.write_trace_events(
+                tensor_name=tensor_name,
```
[the remark this answers is not in the document]
Comment (vandanavk) [reply smdebug/core/hook.py]: changed to training_phase

File: smdebug/core/locations.py (hunk 1 above)
Comment (vandanavk) [reply smdebug/core/locations.py]: changed to new file path format

File: smdebug/core/writer.py (hunk 2 above)
Comment (vandanavk) [reply smdebug/core/writer.py]: not relevant in latest commit

File: smdebug/core/tfevent/timeline_file_writer.py (hunk 3 above)
Comment (vandanavk) [reply smdebug/core/tfevent/timeline_file_writer.py]: not relevant in latest commit

File: smdebug/core/tfevent/timeline_file_writer.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            ev_writer=self._ev_writer,
+            flush_secs=self._flush_secs,
+            sentinel_event=self._sentinel_event,
+        )
+        self._worker.start()
+
+    def write_trace_events(
+        self, training_phase="", op_name="", phase="X", timestamp=None, duration=1, args=None
+    ):
+        duration_in_us = int(duration * 100000)  # convert to micro seconds
```
Comment (NRauschmayr) [inline smdebug/core/tfevent/timeline_file_writer.py:70]: the number is missing a zero: 1.000.000 instead of 100.000

Comment (vandanavk) [reply smdebug/core/tfevent/timeline_file_writer.py:70]: good catch!

Comment (vandanavk) [reply smdebug/core/tfevent/timeline_file_writer.py]: fixed in the latest commit

File: smdebug/core/tfevent/timeline_writer.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        :param args: other information to be added as args
+        :param timestamp: start_time for the event
+        :param duration: any duration manually computed (in seconds)
+        """
+        self.training_phase = training_phase
+        self.phase = phase
+        self.op_name = operator_name
+        self.args = args
+        self.ts_micros = (
+            int(timestamp * 1000000) if timestamp else int(round(time.time() * 1000000))
```
Comment (Vikas-kum) [inline smdebug/core/tfevent/timeline_writer.py:36]: why else part, , timestamp should be mandatory.

File: smdebug/core/tfevent/timeline_writer.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        )
+        self.duration = duration
```
Comment (Vikas-kum) [inline smdebug/core/tfevent/timeline_writer.py:38]: you should check if event is 'X' and duration is not provided then calculate duration here. by doing now() - event_ts

**Comment by vandanavk** [comment]
> i will keep reviewing.
> some initial comments -
> Also, please take care of after file is rotated or new file is created, all the metadata events from previous file should be written in the new file. This means that every file is complete in itself and can be rendered by chrome.
> Please do test scenario for rotation by time and size. you can adjust the environment variable value to speed up testing.
> Also test the hourly rotation, so write a test case where one event is say 12:59 and another event is 1:02 , and check if 2 file is created.
> it would be good if you can update description after testing these cases and also add test cases to validate these.

test for rotation policy is at https://github.com/awslabs/sagemaker-debugger/pull/247/files#diff-77857cca5c5d196c7dc4ff1fe36079acR140. adding all test results in the PR description

so when metadata is written in the new file, is start_time_from_epoch_in_micros the same as the first file or is it updated for this file?

will add a test for hourly rotation

File: smdebug/core/tfevent/timeline_writer.py (hunk 7 above)
Comment (vandanavk) [reply smdebug/core/tfevent/timeline_writer.py:36]: in case user forgets?
I can make it mandatory

**Comment by Vikas-kum** [comment]
> > i will keep reviewing.
> > some initial comments -
> > Also, please take care of after file is rotated or new file is created, all the metadata events from previous file should be written in the new file. This means that every file is complete in itself and can be rendered by chrome.
> > Please do test scenario for rotation by time and size. you can adjust the environment variable value to speed up testing.
> > Also test the hourly rotation, so write a test case where one event is say 12:59 and another event is 1:02 , and check if 2 file is created.
> > it would be good if you can update description after testing these cases and also add test cases to validate these.
> 
> test for rotation policy is at https://github.com/awslabs/sagemaker-debugger/pull/247/files#diff-77857cca5c5d196c7dc4ff1fe36079acR140. adding all test results in the PR description
> 
> so when metadata is written in the new file, is start_time_from_epoch_in_micros the same as the first file or is it updated for this file?
> 
> will add a test for hourly rotation

Right.

File: smdebug/core/locations.py [hunk 9]
```diff
@@ -105,6 +106,30 @@ def get_file_location(self, base_dir=""):
...
+        env_base_location = os.getenv("ENV_BASE_FOLDER", base_dir)
+        date_hour = time.strftime("%y%m%d%H")
+        timestamp = int(round(time.time()))
+        worker_id = get_tb_worker()
+        file_path = os.path.join(
+            env_base_location,
+            "framework/pevents/"
+            + date_hour
+            + "/"
+            + str(timestamp)
```
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply smdebug/core/locations.py:125]: Actually, this is true only if your event.event_ts recorded from python is time.time() . 
Also, if event_ts is time.time(), then your base_start_Time should be 0. 

In reader, we calculate actual time of event by : 
base_start_time_Since_epoch + event_ts ==>  event_ts is relative. 

In your case if event_ts is time.time() , then your base_start_time_since_epoch needs to be 0.

[the remark this answers is not in the document]
Comment (vandanavk) [reply smdebug/core/locations.py:125]: event_ts is relative here. while writing, event_ts = actual_event_time_from_epoch - start_time_from_epoch

[the remark this answers is not in the document]
Comment (vandanavk) [reply smdebug/core/locations.py:125]: also, the file is created lazily. only when maybe_get_timeline_file_Writer() is called from _write_trace_event_summary which is called when event is recorded

**Pull request closed without merging**